### ISSUE TITLE
Support for GCC Builtins

### ISSUE THREAD

**Issue by mrigger** [elliotchance/c2go#778]
We are currently studying the usage of [GCC builtins](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html) by GitHub projects and how well various tools support them. To that end, we are also developing a test suite for the most frequently used machine-independent GCC builtins (available at https://github.com/gcc-builtins/tests). We'd be glad to contribute the test suite (or parts of it) to c2go, if you are interested.

We initially tested c2go and none of the test cases passed, even though I found [25 builtin implementations in the source code](https://github.com/elliotchance/c2go/blob/3498e1a09eb8918756299ef4bc54866e56bc84c4/program/function_definition.go). The reason was that the `assert` from `assert.h` caused an error. From the source code, it seems that [it should be supported](https://github.com/elliotchance/c2go/blob/3498e1a09eb8918756299ef4bc54866e56bc84c4/program/function_definition.go#L66) by c2go, but since `assert` is implemented as a macro, it probably resolves to something incompatible on my Ubuntu system. This is the error I get when using `assert` in a program:

```c
panic: cannot do unary on: &ast.ConditionalOperator{Addr:0x556a1b6d09c8, Pos:ast.Position{File:"/home/manuel/phd/projects/test.c", Line:6, LineEnd:0, Column:21, ColumnEnd:68, StringValue:""}, Type:"int", ChildNodes:[]ast.Node{(*ast.ParenExpr)(0xc42056e6c0), (*ast.IntegerLiteral)(0xc42022c680), (*ast.IntegerLiteral)(0xc42022c700)}}

goroutine 1 [running]:
github.com/elliotchance/c2go/transpiler.transpileUnaryExprOrTypeTraitExpr(0xc4205126e0, 0xc4202f8000, 0xc4206f2038, 0xc4206f2050, 0x40, 0x40, 0x66e380, 0x20, 0xc4206f1fc8, 0x410047, ...)
	/home/manuel/go/src/github.com/elliotchance/c2go/transpiler/unary.go:711 +0x48c
github.com/elliotchance/c2go/transpiler.transpileToExpr(0x7f9120, 0xc4205126e0, 0xc4202f8000, 0x7f9e04aab600, 0x0, 0x0, 0x0, 0x0, 0x82fd68, 0x0, ...)
	/home/manuel/go/src/github.com/elliotchance/c2go/transpiler/transpiler.go:198 +0x2ca
```

Next, we tested a code fragment that uses `__builtin_expect`, which is the most commonly used GCC builtin according to our study:

```c
#include <stdio.h>

int main() {
    volatile void *ptr = NULL;
    if (__builtin_expect(ptr == NULL, 0)) {
        puts("asdf");
    }
}
```

When compiling and executing the fragment, I got the following error:

```
# command-line-arguments
./test.go:200: undefined: __builtin_expect
```

From the source code, it seems that `__builtin_expect` is only supported when `assert.h` is included. In fact, including `assert.h` fixes the error and the program runs as expected after being translated to go. Does this have a special reason? Programs can use this builtin also without including this header file.

Do you have any suggestions on how to proceed with testing the builtin support of c2go?

**Comment by elliotchance** [comment]
Hi @mrigger,

TL;DR:

1. You ran into a bug, if you find a fix and put in a PR that would be awesome.
2. Many features are still not supported, especially if they exist outside of ANSI clang on mac and linux.
3. c2go welcomes all PRs as long as their intention is to support logic as a compiler would. Covering or stubbing header logic or behaviour should only be done if there is no other reasonable choice.

---

c2go always welcomes PRs to support syntax and built-ins. However the main focus has only been "what clang supports" because we must rely on the clang AST to do the transpiling. GCC specific and many generic built-ins have simply not been uncovered yet. Either on platforms outside of those supported on Travis or in code that c2go has not seen yet.

You have definitely run in to a bug above where assert is implemented slightly different with your environment, if you could put in a fix that would be fantastic. But keep in mind...

I want c2go to be as low-level as possible. Properly parsing headers and macros just as the compiler would and avoid stubbing or otherwise covering logic because it's neater or more generic. It must be able to handle C files warts and all, just as the many C compilers have to do.

That being said, it's common for compilers to treat some objects normally defined in headers as always available. If people have go used to that practice that effetely puts that into the "features" of that compiler. If it's common practice to do it this way (for example using asserts without including the proper header) then it makes sense to extend this to c2go as long as it doesn't conflict with the previous paragraph.

**Comment by mrigger** [comment]
Thank you for the detailed answer! I'll check if I can fix the bug under these constraints.

Compiler builtins are actually supposed to always be available (if supported), even if no header files are included. GCC, Clang, and ICC all support the `__builtin_expect` test case above. I'm not sure if I understand why c2go only supports builtins if certain header files are included. Can the existing ones not made to always be available?

**Comment by elliotchance** [comment]
If that is the case then c2go should do the same thing. The ultimate goal is to have c2go as drop in replacement for a general C compiler so it should follow and behave the way that most other compilers do.
